cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+          lateral_averaging_properties.push_back(std::make_unique<internal::FunctorDepthAverageUnscaledViscosity<dim>>());
+
+          std::vector<std::vector<double>> averages =
+            this->get_lateral_averaging().compute_lateral_averages(reference_viscosity_coordinates,
+                                                                   lateral_averaging_properties);
+
+          average_viscosity_profile.swap(averages[0]);
+
+          for (const auto &lateral_viscosity_average: average_viscosity_profile)
+            AssertThrow(numbers::is_finite(lateral_viscosity_average),
+                        ExcMessage("In computing depth averages, there is at"
+                                   " least one depth band that does not have"
+                                   " any quadrature points in it."
+                                   " Consider reducing number of depth layers"
+                                   " for averaging."));
+        }
+
+      initialized = true;
+    }
+
+
+
+    template <int dim>
+    std::pair<double, unsigned int>
+    EquilibriumGrainSize<dim>::get_reference_viscosity (const double depth) const
+    {
+      // Make maximal depth slightly larger to ensure depth < maximal_depth
+      const double maximal_depth = this->get_geometry_model().maximal_depth() *
+                                   (1.0+std::numeric_limits<double>::epsilon());
+      (void) maximal_depth;
+
+      Assert(depth < maximal_depth, ExcInternalError());
+
+      unsigned int depth_index;
+      if (depth < reference_viscosity_coordinates.front())
+        {
+          depth_index = 0;
+        }
+      else if (depth > reference_viscosity_coordinates.back())
+        {
+          depth_index = reference_viscosity_coordinates.size() - 1;
+        }
+      else
+        {
+          depth_index = std::distance(reference_viscosity_coordinates.begin(),
+                                      std::lower_bound(reference_viscosity_coordinates.begin(),
+                                                       reference_viscosity_coordinates.end(),
+                                                       depth));
+        }
+      if (depth_index > 0)
+        depth_index -= 1;
+
+      // When evaluating reference viscosity, evaluate at the next lower depth that is stored
+      // in the reference profile instead of the actual depth. This makes the profile piecewise
+      // constant. This will be specific to the viscosity profile used (and ignore the entry with
+      // the largest depth in the profile).
+      double reference_viscosity = reference_viscosity_profile->compute_viscosity(reference_viscosity_coordinates.at(depth_index));
+
+      // This parameter is only because we change the asthenosphere viscosity in our models.
+      // By default, it is set to the value in the reference profile.
+      if (depth_index == 1)
+        reference_viscosity = asthenosphere_viscosity;
```

Comment:
What does "this parameter" refer to?

Reply:
It is the `asthenosphere_viscosity`. I will remove the comment, it is not relevant anymore.